### PR TITLE
use cluster api patch helper to do the scope close

### PR DESCRIPTION
**Pull request by xrmzju**
**What this PR does / why we need it**:
fix issue: https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/1019
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**Release note**:

```release-note

```

### CODE REVIEW

**Comment by vincepri** [comment]
@xrmzju It seems that the issue can be fixed by issuing Status().Patch() before the normal Patch(), see https://github.com/kubernetes-sigs/cluster-api/pull/1261

**Comment by xrmzju** [comment]
I signed it

**Comment by xrmzju** [comment]
> @xrmzju It seems that the issue can be fixed by issuing Status().Patch() before the normal Patch(), see [kubernetes-sigs/cluster-api#1261](https://github.com/kubernetes-sigs/cluster-api/pull/1261)

this is my first solution, and i found a problem here. i  will generate certs info  in my provider controller if not specified in spec and store it in a secret, and after first reconcile,  the certs info and secret are generated normally, after status.Patch, it will trigger a new update event, and the  cert info in the provider spec is still  nil<status.Patch does not update spec>, so the second reconcile round would do the generate certs action again without update the secret info(cause i'll try to get the secret, if exist i will not create it), that causes inconsistence between the cert info in the spec and stored in the secret.                    
the above case is only to declare that if we try to do some <one-time action > in our logic, just put status.Patch before the client.Patch will cause some problem. i think the status copy action would be the most secure one.

**Comment by vincepri** [comment]
@xrmzju Thanks for the detailed explanation! The change LGTM, can we add unit tests to cover the behaviors you've described? It'd be great to cover both the failing and success cases with DeepCopy, so we can track this issue.

@detiber is also working on a generic implementation of patching in Cluster API, once that's ready we should switch to that one.

**Comment by ncdc** [comment]
@detiber your patch helper will handle this, won't it?

**Comment by detiber** [comment]
> @detiber your patch helper will handle this, won't it?

That is correct, an example of usage can be found here: https://github.com/kubernetes-sigs/cluster-api-bootstrap-provider-kubeadm/pull/148

**Comment by ncdc** [comment]
@xrmzju do you want to adjust this PR to be based on the new patch helper & the example Jason provided? If not, please let us know, and we can get someone else to work on it.

**Comment by xrmzju** [comment]
> @xrmzju do you want to adjust this PR to be based on the new patch helper & the example Jason provided? If not, please let us know, and we can get someone else to work on it.
@ncdc i'll adjust it.

File: pkg/cloud/scope/machine.go [hunk 1]
```diff
@@ -164,15 +167,5 @@ func (m *MachineScope) SetAnnotation(key, value string) {
...
-	if err := m.client.Patch(ctx, m.AWSMachine, m.patch); err != nil {
-		return errors.WithStack(err)
-	}
-
-	if err := m.client.Status().Patch(ctx, m.AWSMachine, m.patch); err != nil {
-		return errors.WithStack(err)
-	}
-
-	return nil
+	return m.patchHelper.Patch(context.Background(), m.Machine)
```
Comment (detiber) [inline pkg/cloud/scope/machine.go]:
```suggestion
	return m.patchHelper.Patch(context.Background(), m.AWSMachine)
```

File: pkg/cloud/scope/cluster.go [hunk 2]
```diff
@@ -66,13 +67,17 @@ func NewClusterScope(params ClusterScopeParams) (*ClusterScope, error) {
 		params.AWSClients.ELB = elb.New(session)
 	}
 
+	helper, err := patch.NewHelper(params.AWSCluster, params.Client)
+	if err != nil {
+		return nil, errors.Errorf("failed to init patch helper: %v", err)
```
Comment (ncdc) [inline pkg/cloud/scope/cluster.go]: Please use `errors.Wrap`

File: pkg/cloud/scope/machine.go [hunk 3]
```diff
@@ -63,25 +63,28 @@ func NewMachineScope(params MachineScopeParams) (*MachineScope, error) {
 		params.Logger = klogr.New()
 	}
 
+	helper, err := patch.NewHelper(params.AWSMachine, params.Client)
+	if err != nil {
+		return nil, errors.Errorf("failed to init patch helper: %v", err)
```
Comment (ncdc) [inline pkg/cloud/scope/machine.go]: Please use `errors.Wrap`

File: pkg/cloud/scope/machine.go [hunk 4]
```diff
@@ -63,25 +63,28 @@ func NewMachineScope(params MachineScopeParams) (*MachineScope, error) {
...
+		AWSMachine:  params.AWSMachine,
+		Logger:      params.Logger,
+		patchHelper: helper,
 	}, nil
 }
 
 // MachineScope defines a scope defined around a machine and its cluster.
 type MachineScope struct {
 	logr.Logger
-	patch  client.Patch
```
Comment (ncdc) [inline pkg/cloud/scope/machine.go]: A lot of times we separate the unexported from the exported fields, as was the case here. I'd recommend you keep this pattern - move `patchHelper` back up here, and preserve the blank line in between `client` and `Cluster`.

File: pkg/cloud/scope/machine.go (hunk 1 above)
Comment (xrmzju) [reply pkg/cloud/scope/machine.go]: my fault... done.

File: pkg/cloud/scope/cluster.go (hunk 2 above)
Comment (xrmzju) [reply pkg/cloud/scope/cluster.go]: done

Comment (xrmzju) [reply pkg/cloud/scope/cluster.go]: done

File: pkg/cloud/scope/machine.go (hunk 3 above)
Comment (xrmzju) [reply pkg/cloud/scope/machine.go]: done

File: pkg/cloud/scope/machine.go (hunk 4 above)
Comment (xrmzju) [reply pkg/cloud/scope/machine.go]: ok, i'll keep that pattern

**Comment by ncdc** [comment]
/lgtm
/approve

/assign @detiber 
Jason, can you give it one final look, then cancel the hold? Thanks!

**Pull request merged**